File: clients/ruby/commons/lib/api_gouv_commons/response.rb
module ApiGouvCommons
  class Response
    attr_reader :raw, :http_status, :headers, :rate_limit

    def initialize(raw:, http_status:, headers:, rate_limit: nil)
      @raw = raw.is_a?(Hash) ? raw : {}
      @http_status = http_status
      @headers = headers || {}
      @rate_limit = rate_limit
    end

    def data
      raw['data']
    end

    def links
      raw['links'] || {}
    end

    def meta
      raw['meta'] || {}
    end

    def success?
      http_status.to_i.between?(200, 299)
    end
  end
end

File: clients/node/commons/src/response.ts
import { RateLimit } from './rate-limit.js';

/** Parsed API response with envelope fields (data, links, meta) and rate limit info. */
export class Response {
  readonly raw: Record<string, unknown>;
  readonly httpStatus: number;
  readonly headers: Record<string, string>;
  readonly rateLimit: RateLimit | null;

  constructor(options: {
    raw: unknown;
    httpStatus: number;
    headers: Record<string, string>;
    rateLimit?: RateLimit | null;
  }) {
    this.raw = isRecord(options.raw) ? options.raw : {};
    this.httpStatus = options.httpStatus;
    this.headers = options.headers;
    this.rateLimit = options.rateLimit ?? null;
  }

  get data(): unknown {
    return this.raw['data'];
  }

  get links(): Record<string, unknown> {
    return (this.raw['links'] as Record<string, unknown>) ?? {};
  }

  get meta(): Record<string, unknown> {
    return (this.raw['meta'] as Record<string, unknown>) ?? {};
  }

  get success(): boolean {
    return this.httpStatus >= 200 && this.httpStatus < 300;
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
